src/_pytest/timing.py

```diff
@@ -20,6 +21,51 @@
     from pytest import MonkeyPatch
 
 
+@dataclasses.dataclass(frozen=True)
+class Instant:
+    """
+    Measures timing between different points in the code.
+
+    Inspired by Rust's `std::time::Instant`.
+    """
+
+    # Creation time of this instant, using time.time(), to measure actual time.
+    # Use a `lambda` to initialize the default to correctly get the mocked time via `MockTiming`.
+    _start_time: float = dataclasses.field(default_factory=lambda: time())
+
+    # Initial "tick" of the performance counter to measure precise elapsed time.
+    # Use a `lambda` to initialize the default to correctly get the mocked time via `MockTiming`.
+    _start_perf: float = dataclasses.field(default_factory=lambda: perf_counter())
+
+    def duration(self) -> Duration:
+        """Measure the duration since `Instant` was created."""
+        return Duration(
+            start=self._start_time,
+            elapsed_s=perf_counter() - self._start_perf,
+            stop=time(),
+        )
+
+
+@dataclasses.dataclass(frozen=True)
+class Duration:
+    """A slice of time as measured by `Instant.duration()`."""
+
+    # Start time of the duration, as seconds since epoch.
+    start: float
+    # Stop time of the duration, as seconds since epoch.
+    stop: float
+    # Elapsed time of the duration, in seconds, measured using a performance counter for precise timing.
+    elapsed_s: float
+
+    def start_utc(self) -> datetime:
+        """Start time of the duration, as a datetime in UTC."""
+        return datetime.fromtimestamp(self.start, timezone.utc)
+
```

Comment:
Id prefer an as_utc method on instant 

That requires durations to be in terms of those instants

Reply:
I think I get now the API you intend, is it something like this (roughly)?

```python
class Instant:
    # time using time.time()
    time: float
    # used by elapsed_from() to compute elapsed time from another instant
    _perf_counter: float

    def elapsed_from(self, other: Instant) -> float: ... # elapsed time in seconds, using _perf_counter

    def as_utc(self) -> datetime: ... 
```

The usage of this API in `runner.py` would then be something like:

```python
start = Instant()
...
end = Instant()
duration = end.elapsed_from(end)
```

Is that correct?

Reply:
i'd have a helper method in the instant to get a duration that starts at that instant and ens add a new instant

Reply:
Or perhaps you just want to change `Duration.start` and `Duration.end` to `Instant` instances instead of `float`?

Reply:
@RonnyPfannschmidt please take a look at my last commit. 👍